src/Orleans/Async/TaskExtensions.cs

```diff
@@ -216,7 +216,7 @@ internal static Task<T> ConvertTaskViaTcs<T>(Task<T> task)
                     {
                         resolver.TrySetResult(t.Result);
                     }
-                });
+                }, TaskContinuationOptions.ExecuteSynchronously);
```

Comment:
The `InvokeMethodAsync` method can become the following, which should require less allocations (no async state machine) and preserves the stack on faults/cancellations.
```c#
protected Task<T> InvokeMethodAsync<T>(int methodId, object[] arguments, InvokeMethodOptions options = InvokeMethodOptions.None, SiloAddress silo = null)
{
    object[] argsDeepCopy = null;
    if (arguments != null)
    {
        CheckForGrainArguments(arguments);
        argsDeepCopy = (object[])SerializationManager.DeepCopy(arguments);
    }
            
    var request = new InvokeMethodRequest(this.InterfaceId, methodId, argsDeepCopy);

    if (IsUnordered)
        options |= InvokeMethodOptions.Unordered;

    Task<object> resultTask = InvokeMethod_Impl(request, null, options);

    if (resultTask == null)
    {
        return Task.FromResult(default(T));
    }

    if (resultTask.Status == TaskStatus.Created)
    {
        resultTask.Start();
    }

    return resultTask.ContinueWith(
        _ => (T)_.Result,
        TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously);
}
```

What do you think?